More MPU work

kjbracey opened the pull request:
### Description

* Don't call HAL MPU free code when `platform.use-mpu` is turned off.
* Correct MPU synchronisation sequences.
* Reduce MPU code size further with a loop
* Use higher-level calls, fix ARMv8-M error 
* Assert number of regions, rather than error
* ARMv7-M: correctly protect 80000000 RAM region
* Fix target.mpu-rom-end setting, for ARMv8-M too
* nRF52840: Set mpu-rom-end to 0x1fffffff (to test this case)

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

kjbracey committed:
Fix MPU synchronisation

Synchronisation instructions were not quite right - too strict on entry,
and not quite correctly synchronising the instruction stream on exit.

References:

* https://static.docs.arm.com/dai0321/a/DAI0321A_programming_guide_memory_barriers_for_m_profile.pdf
* https://static.docs.arm.com/100699/0100/armv8m_architecture_memory_protection_unit_100699_0100_00_en.pdf

c1728p9 committed:
Use MPU platform deinit rather than hal free

Use the MPU through the platform layer rather than through the HAL.

kjbracey committed:
Reduce dynamic MPU code size with a loop

kjbracey committed:
MPU - use higher-level calls, fix ARMv8-M error

Switch to higher-level calls and macros, and fix an error in the ARMv8-M
version - "inner" attributes were not being set correctly due to a
copy/paste error - "outer" was being set twice.

This means RAM would have been marked WTRA rather than WBWA for the
inner cache.

Slightly reduces ARMv7-M init code size by feeding region number
into RBAR instead of using RNR.

kjbracey commented:
Needs some more local testing - would be good if @c1728p9 could have a look at ARMv8-M in particular. Also intending one more commit to change MBED_ERRORs to MBED_ASSERTs.

cmonr commented:
@kjbracey-arm Don't forget the astyle fix :)

kjbracey committed:
Assert MPU regions, rather than error

As we build for a specific CPU, a runtime check for number of MPU
regions in release builds is not worthwhile. Make it an assert only.
Saves a little space in develop images, a lot in release.

kjbracey commented:
Adjusted a bit more, making ARMv7-M use higher-level API too.

ARMv7-M init code now down to 170 bytes from 215.

0xc0170 commented:
Please review astyle travis CI

kjbracey commented:
Full mbed_mpu_v7m.c + mbed_mpu_mgmt.c is now 246+351 = 597, down from 659 in current master.

kjbracey committed:
ARMv7-M MPU: correctly protect 80000000 RAM region

Subregion mask for this region was set fully disabled, instead of fully
enabled.

kjbracey committed:
Fix target.mpu-rom-end setting, for ARMv8-M too

targets.json was not specifying the same macro name as the code was
checking for, so setting was ineffective.

Making this work tripped up not-supported checks in ARMv8-M - rather than deal
with making this work, support it instead.

Both ARMv7-M and ARMv8-M slightly reduce code size and runtime impact if
mpu-rom-end is 0x1fffffff, using one fewer region.

This means default setup for ARMv8-M now requires 5 regions, with
mpu-rom-end set to default 0x0fffffff, but this can be put back to 4 by
changing the setting.

kjbracey committed:
nRF52840: Set mpu-rom-end to 0x1fffffff

So we have at least one test platform exercising the special case of
mpu-rom-end being 0x1fffffff, set that for nRF52840.

kjbracey commented:
Two more bugs found - `target.mpu-rom-end` JSON setting wasn't working, and 80000000 wasn't protected in ARMv7-M.

Fixed up, including making mpu-rom-end work on ARMv8-M - was simpler than figuring out what to do to not hit the "not supported" test once JSON was fixed.

Also will save a few more bytes if you do set mpu-rom-end to 0x1fffffff.

cmonr commented:
CI started

kjbracey commented:
This does really need rereview - I did some more significant stuff since @c1728p9 last looked at it.

cmonr commented:
Hmm... I wonder.

I recall a time when GitHub would auto-dismiss reviews when a change was made.

c1728p9 committed:
Update MPU ram/rom split for Nuvoton M2351

Move the start of RAM from 0x10000000 to 0x20000000 on the
NUMAKER_PFM_M2351. This allows the target to work correctly.

c1728p9 reviewed (approved):
I tested this on a `NUMAKER_PFM_M2351` and got crashes on boot due to the MPU. I pushed a commit to update the ram/rom split for this target to correct the crashes. This PR looks good to me. Thanks @kjbracey-arm!

0xc0170 commented:
CI started
